File: python_sdk/src/reality_capture/specifications/eval_o3d.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class EvalO3DInputs(BaseModel):
    reference: str = Field(description="Reality data id of ContextScene, annotated with embedded 3D object references")
    prediction: str = Field(description="Reality data id of ContextScene, "
                                        "annotated with embedded 3D object predictions")


class EvalO3DOutputs(BaseModel):
    report: str = Field(description="Reality data id of json report with binary classification")
    objects3d: str = Field(alias="objects3D", description="Reality data id of ContextScene, "
                                                          "annotated with classified embedded 3D objects")


class EvalO3DOutputsCreate(Enum):
    REPORT = "report"
    OBJECTS3D = "objects3D"


class EvalO3DOptions(BaseModel):
    threshold_iou: Optional[int] = Field(None, alias="thresholdIOU", description="Intersection over union threshold")


class EvalO3DSpecificationsCreate(BaseModel):
    inputs: EvalO3DInputs = Field(description="Inputs")
    outputs: list[EvalO3DOutputsCreate] = Field(description="Outputs")
    options: Optional[EvalO3DOptions] = Field(None, description="Options")


class EvalO3DSpecifications(BaseModel):
    inputs: EvalO3DInputs = Field(description="Inputs")
    outputs: EvalO3DOutputs = Field(description="Outputs")
    options: Optional[EvalO3DOptions] = Field(None, description="Options")

File: typescript/packages/reality-capture/src/specifications/eval_o3d.ts
import { z } from "zod";

export const EvalO3DInputsSchema = z.object({
  reference: z.string().describe("Reality data id of ContextScene, annotated with embedded 3D object references"),
  prediction: z.string().describe("Reality data id of ContextScene, annotated with embedded 3D object predictions"),
});
export type EvalO3DInputs = z.infer<typeof EvalO3DInputsSchema>;

export const EvalO3DOutputsSchema = z.object({
  report: z
    .string()
    .regex(/^bkt:.+/)
    .describe("Path in Bucket of json report with binary classification")
    .optional(),
  objects3d: z
    .string()
    .describe("Reality data id of ContextScene, annotated with classified embedded 3D objects")
    .optional(),
});
export type EvalO3DOutputs = z.infer<typeof EvalO3DOutputsSchema>;

export enum EvalO3DOutputsCreate {
  REPORT = "report",
  OBJECTS3D = "objects3D",
}

export const EvalO3DOptionsSchema = z.object({
  thresholdIOU: z
    .number()
    .describe("Intersection over union threshold")
    .optional(),
});
export type EvalO3DOptions = z.infer<typeof EvalO3DOptionsSchema>;

export const EvalO3DSpecificationsCreateSchema = z.object({
  inputs: EvalO3DInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(EvalO3DOutputsCreate)).describe("Outputs"),
  options: EvalO3DOptionsSchema.describe("Options").optional(),
});
export type EvalO3DSpecificationsCreate = z.infer<typeof EvalO3DSpecificationsCreateSchema>;

export const EvalO3DSpecificationsSchema = z.object({
  inputs: EvalO3DInputsSchema.describe("Inputs"),
  outputs: EvalO3DOutputsSchema.describe("Outputs"),
  options: EvalO3DOptionsSchema.describe("Options").optional(),
});
export type EvalO3DSpecifications = z.infer<typeof EvalO3DSpecificationsSchema>;